feat: update FindPriorBackups to use backup index

kev-cao opened the pull request:
This patch teaches `FindPriorBackups` (now more aptly renamed to `FindAllIncrementals`) to use the backup index introduced in #150384. Additional changes will still be required to fully fix the related tickets, namely updating the logic to no longer need to load ALL manifests of a chain into memory, only the manifests needed.

Epic: CRDB-47942

Informs: #150327, #150329, #150330, #150331

jeffswenson commented on pkg/backup/backupdest/backup_destination_test.go:
```diff
@@ -351,47 +347,6 @@ func TestBackupRestoreResolveDestination(t *testing.T) {
 					// We also wrote a new full backup, so let's update the latest.
 					writeLatest(t, collectionLoc, expectedSuffix)
 				}
-
```
Is there a test that covers the case that was removed here?

jeffswenson commented on pkg/backup/backupdest/backup_index.go:
```diff
@@ -31,11 +35,25 @@ import (
 	details jobspb.BackupDetails,
 ) error {
-	ctx, sp := tracing.ChildSpan(ctx, "backupdest.WriteBackupIndexMetadata")
+	indexStore, err := makeExternalStorageFromURI(
```
I think we are leaking this cloud storage instance and we need to defer a Close.

jeffswenson commented on pkg/backup/backup_test.go:
```diff
@@ -718,7 +718,6 @@ func TestBackupAndRestoreJobDescription(t *testing.T) {
 	}
 
 	sqlDB.Exec(t, "BACKUP INTO ($1, $2, $3)", collections...)
-	sqlDB.Exec(t, "BACKUP INTO LATEST IN ($1, $2, $3)", collections...)
```
Why was this removed from the test?

kev-cao replied:
I removed it mainly for simplicity's sake. The test just checks the description outputted once the backup statement is fully resolved, so in this case, `INTO LATEST` just checks that `LATEST` is resolved accordingly.

We do already have a `BACKUP INTO LATEST ... WITH incremental_location` case which covers `LATEST`. If we wanted to still have them tested separately, it just would require a decent amount of changes to the test since you'd have to also update the logic for finding the full path that `LATEST` gets resolved into and separate the two full paths.

jeffswenson commented on pkg/backup/backupdest/backup_destination.go:
```diff
@@ -237,41 +241,40 @@ func ResolveDest(
-		partName = partName + "-" + startTime.GoTime().Format(backupbase.DateBasedIncFolderNameSuffix)
 	}
+
+	partName := ConstructDateBasedIncrementalFolderName(startTime.GoTime(), endTime.GoTime())
```
Is it safe to use the new partName format even in the case where we are not writing to the backup index?

jeffswenson reviewed (commented):
Looks good overall 🥳

kev-cao commented on pkg/backup/backupinfo/manifest_handling.go:
```diff
@@ -1000,11 +1001,38 @@ func validateContinuity(manifests []backuppb.BackupManifest) error {
+	// increasing end time and then by increasing start time to break ties. The
+	// latter ensures that when we elide duplicate end times, we keep the most
+	// compacted backup.
+	permutation := make([]int, len(backups))
```
Oh I like that approach a lot, I'll give it a shot. Thanks!

kev-cao committed:
feat: update FindPriorBackups to use backup index

This patch teaches FindPriorBackups (now more aptly renamed to
FindAllIncrementals) to use the backup index introduced in #150384.
Additional changes will still be required to fully fix the related
tickets, namely updating the logic to no longer need to load ALL
manifests of a chain into memory, only the manifests needed.

Epic: CRDB-47942

Informs: #150327, #150329, #150330, #150331

Release note: None

msbutler commented on pkg/backup/backupdest/backup_index.go:
```diff
@@ -138,6 +139,84 @@ func IndexExists(ctx context.Context, store cloud.ExternalStorage, subdir string
+// Note: While the path is stored in the index file, we can take a shortcut here
+// and derive it from the filename solely because backup paths are
+// millisecond-precise and so are the timestamps encoded in the filename.
+func parseBackupFilePathFromIndexFileName(subdir, basename string) (string, error) {
```
I thought we were going to store the path to the full in the index file, so why do we need to parse the index file path?

kev-cao replied:
I was thinking that even though we do store the path in the index file, since we could actually parse it from the filename, we may as well do so to save us some reads.

And the full path would still be stored there as our source of truth in case we do ever mess with the filenames.

msbutler replied:
> we may as well do so to save us some reads
why do you think the cost of reading the index file will be expensive? IMHO the "cost" of maintaining this parsing logic significantly outweighs the "cost" loading the index file (which is tiny!)

msbutler replied:
restore/backup will be slow because they gotta read and write a whole cluster to external storage, not because they have to read a little metadata file ;)

kev-cao replied:
Yeah I'm falling for that premature optimization here — I was thinking of trying to keep it performant at the helper function level in the event that some other area of the codebase would use it. But since no other place actually does use it, I'm probably getting ahead of myself here.

msbutler commented on pkg/backup/backupdest/backup_destination.go:
```diff
@@ -237,41 +237,40 @@ func ResolveDest(
 
-	// Within the chosenSuffix dir, differentiate incremental backups with partName.
-	partName := endTime.GoTime().Format(backupbase.DateBasedIncFolderName)
-	if execCfg.Settings.Version.IsActive(ctx, clusterversion.V25_2) {
```
nit: looks like you're removing a version gate in this part of the commit. I think this should be a separate commit.

msbutler reviewed (commented):
flushing comments from first commit, as they may lead to some significant refactoring in the second commit :D. Also, I think this PR could be rebased on master (to remove the `TestBackupRestoreSeparateExplicitIsDefault ` diff)

kev-cao commented on pkg/backup/backuputils/utils.go:
```diff
@@ -102,13 +112,15 @@ func EncodeDescendingTS(ts time.Time) string {
 //	collectionURI: "nodelocal://1/collection"
-//	backupURI: "nodelocal://1/collection/backup1/"
-//	returns: "backup1/"
+//	backupURI: "nodelocal://1/collection/subdir/inc/"
+//	returns: "subdir/inc/"
```
TODO: In a follow up PR, write tests to assert these sorts of guarantees.

kev-cao commented on pkg/backup/backupdest/incrementals.go:
Yeah that sounds good to me. I don't think it necessarily applies here, but what do you think of this [coding guideline](https://cockroachlabs.atlassian.net/wiki/spaces/CRDB/pages/181371303/Go+Golang+coding+guidelines#Try-to-avoid-bool-parameters) in our docs?

kev-cao commented on pkg/backup/backupdest/backup_destination.go:
```diff
@@ -537,10 +544,13 @@ func ListFullBackupsInCollection(
 // included in the result, otherwise they are filtered out.
 func ResolveBackupManifests(
 	ctx context.Context,
+	execCfg *sql.ExecutorConfig,
```
Yeah it is absolutely horrifying to look at, especially if you're being onboarded onto this. I remember looking at this a year ago and having an aneurysm

kev-cao commented on pkg/backup/compaction_job.go:
```diff
@@ -721,8 +721,18 @@ func getBackupChain(
+		return nil, nil, nil, nil, err
+	}
 	resolvedBaseDirs, resolvedIncDirs, _, err := resolveBackupDirs(
-		ctx, execCfg, user, dest.To, dest.IncrementalStorage, dest.Subdir,
```
It was just because the default URI and a resolved subdir were both parameters that were added to `ResolveBackupManifests`, and since I already resolved the subdir outside of `resolveBackupDirs`, I figured I may as well pass it a resolved subdir to avoid another resolution step.

kev-cao replied:
Oh yeah I guess since `resolveBackupDirs` is only ever called here, then yeah we could put the onus on the caller to resolve subdirs.

kev-cao commented on pkg/backup/restore_planning.go:
```diff
@@ -1686,9 +1686,14 @@ func doRestorePlan(
 	if strings.EqualFold(subdir, backupbase.LatestFileName) {
 		// set subdir to content of latest file
-		latest, err := backupdest.ReadLatestFile(ctx, from[0],
+		latest, err := backupdest.ReadLatestFile(ctx, defaultCollectionURI,
```
Sort of a similar situation with the last comment. I needed to resolve the default collection URI to pass into `ResolveBackupManifests`, so I just used that instead of `from[0]`. Actually, something I did notice was that in certain places we assume that `from[0]` is always the default URI but it's not a guarantee. So that means that if the user were to pass in a default locality but it wasn't the first URI in the list, then we would treat the wrong URI as the default. Of course, there's multiple places where these need fixed, but I'm thinking maybe we do something similar to subdir normalization. We could at the SQL layer in the statement planning ensure that the details always have the first URI be the default.

kev-cao commented on pkg/backup/show.go:
```diff
@@ -336,7 +336,11 @@ you must pass the 'encryption_info_dir' parameter that points to the directory o
 			info        backupInfo
 			memReserved int64
 		)
-		info.collectionURI = dest[0]
```
Same reason as above

msbutler commented on pkg/backup/backupdest/incrementals.go:
```diff
@@ -191,23 +275,25 @@ func ResolveIncrementalsBackupLocation(
-	_, err = backupsFromLocation(ctx, user, execCfg, resolvedIncrementalsBackupLocation[0])
+	_, err = backupsFromLocation(
+		ctx, user, execCfg, defaultCollectionURI, subdir,
+		resolvedIncrementalsBackupLocation[0], false, /* customIncLocation */
```
nit, i'd instead define `customIncLocation := len(explicitIncrementalCollections) > 0 ` above the branch and pass that instead of hard coded bools. That protects against a future bug where someone monkeys around with the flow control here.

msbutler reviewed (commented):
nearly there!

kev-cao commented on pkg/backup/backupdest/incrementals_test.go:
```diff
@@ -124,10 +127,366 @@ func TestJoinURLPath(t *testing.T) {
+	)
+	require.NoError(
+		t,
+		cloud.WriteFile(context.Background(), backupStore, backupbase.BackupManifestName, emptyReader),
```
Actually they write a metadata and a manifest file, needed since some of our logic still depends on one over the other :(

msbutler replied:
💀

msbutler commented on pkg/backup/backupdest/backup_index.go:
```diff
@@ -29,6 +30,15 @@ import (
+		settings.ApplicationLevel,
+		"backup.index.read.enabled",
+		"if true, the backup index will be read when reading from a backup collection",
+		false,
```
if you wrapped the value like so:
```
metamorphic.ConstantWithTestBool("backup.index.read.enabled", false)
```

then our unit tests and non-perf roachtests will randomly use the index. for non-test clusters, this setting will continue to be default false.

msbutler reviewed (approved):
One more big one!

kev-cao commented:
TFTR!

bors r=msbutler
